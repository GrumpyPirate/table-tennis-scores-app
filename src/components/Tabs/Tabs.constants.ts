import { rem } from 'polished';
import styled, { css } from 'styled-components';

import styleConfig from 'config/styles';

// Config
export const tabs = {
  PLAYERS: {
    id: 'PLAYERS',
    label: 'Players',
  },
  GAMES: {
    id: 'GAMES',
    label: 'Games',
  },
};

// Components
export const TabList = styled.ul`
  ${styleConfig.mixins.listReset}
  display: flex;
  margin: 0 auto ${rem(32)};
  border-radius: ${rem(24)};
  background-color: #fff;
  overflow: hidden;
  box-shadow: ${styleConfig.shadows.card};
  max-width: ${rem(480)};
`;

export const TabButton = styled.button<{ isActive: boolean }>`
  ${styleConfig.mixins.buttonReset}
  display: block;
  width: 100%;
  padding: ${rem(8)} ${rem(12)};
  font-weight: bolder;
  background-color: #fff;

  @media ${styleConfig.mediaQueries.largeMobile} {
    display: inline-block;
  }

  ${({ isActive }) =>
    isActive &&
    css`
      background-color: ${styleConfig.palette.greyDark};
      color: #fff;
      border-radius: ${rem(24)};
    `}
`;

export const Tab = styled.li`
  flex: 1 1 50%;

  &::before {
    content: normal;
  }

  &:first-of-type {
    ${TabButton} {
      border-top-left-radius: ${rem(24)};
      border-bottom-left-radius: ${rem(24)};
    }
  }

  &:last-of-type {
    ${TabButton} {
      border-top-right-radius: ${rem(24)};
      border-bottom-right-radius: ${rem(24)};
    }
  }
`;
