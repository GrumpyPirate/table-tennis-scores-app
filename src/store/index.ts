import { configureStore } from '@reduxjs/toolkit';
import { combineReducers } from 'redux';

import gameResults, { initialState as initialGameResultsState } from './slices/gameResults';
import players, { initialState as initialPlayersState } from './slices/players';

/**
 * State slices have been set up using the ðŸ¦† pattern. I.e. reducer as default export, everything else (actions, thunks) as named
 * exports.
 */
export const rootReducer = combineReducers({ players, gameResults });

// Extract the RootState type, so that we can use it across the app
export type RootState = ReturnType<typeof rootReducer>;

// Initial root state, exported so that we can reuse this elsewhere (e.g. tests)
export const initialRootState: RootState = {
  players: initialPlayersState,
  gameResults: initialGameResultsState,
};

const store = configureStore({
  reducer: rootReducer,
  preloadedState: initialRootState,
});

// Extract the configured store's Dispatch type, which gives us typings for the middleware as well
export type AppDispatch = typeof store.dispatch;

export default store;
