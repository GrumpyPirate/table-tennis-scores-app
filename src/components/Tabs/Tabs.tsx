import React, { FunctionComponent, useEffect, useState } from 'react';
import { useDispatch } from 'react-redux';

import GamesView from 'components/GamesView/GamesView';
import PlayersView from 'components/PlayersView/PlayersView';
import { AppDispatch } from 'store';
import { readAllGameResults } from 'store/slices/gameResults';
import { readAllPlayers } from 'store/slices/players';

import { Tab, TabButton, TabList, tabs } from './Tabs.constants';

const Tabs: FunctionComponent = () => {
  const dispatch = useDispatch<AppDispatch>();
  const [activeTab, setActiveTab] = useState(tabs.PLAYERS.id);

  /**
   * Use a side-effect to:
   *  - Fetch the list of all players, after the first render
   */
  useEffect(() => {
    dispatch(readAllPlayers());
    dispatch(readAllGameResults());
  }, [dispatch]);

  return (
    <>
      <nav>
        <TabList>
          {Object.values(tabs).map(({ id, label }) => (
            <Tab key={`tab--${id}`}>
              <TabButton
                type="button"
                onClick={() => {
                  setActiveTab(id);
                }}
                isActive={activeTab === id}
              >
                {label}
              </TabButton>
            </Tab>
          ))}
        </TabList>
      </nav>
      <section>
        {activeTab === tabs.PLAYERS.id && (
          <div data-testid={`tabpanel--${tabs.PLAYERS.id}`}>
            <PlayersView />
          </div>
        )}
        {activeTab === tabs.GAMES.id && (
          <div data-testid={`tabpanel--${tabs.GAMES.id}`}>
            <GamesView />
          </div>
        )}
      </section>
    </>
  );
};

export default Tabs;
