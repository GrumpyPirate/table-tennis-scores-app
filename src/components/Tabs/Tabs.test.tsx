import { fireEvent, render } from '@testing-library/react';
import React from 'react';
import { Provider } from 'react-redux';

import store from '../../store';
import Tabs from './Tabs';
import { tabs } from './Tabs.constants';

describe('Components/Tabs', () => {
  test('should render as expected, without crashing', () => {
    const { container } = render(
      <Provider store={store}>
        <Tabs />
      </Provider>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  test('should display the appropriate tabpanel, when activating each tab', () => {
    const { getAllByText, getByTestId } = render(
      <Provider store={store}>
        <Tabs />
      </Provider>,
    );

    // Activate the Players tab
    fireEvent.click(getAllByText(tabs.PLAYERS.label)[0]);
    expect(getByTestId(`tabpanel--${tabs.PLAYERS.id}`)).toBeTruthy();

    // Activate the Gamess tab
    fireEvent.click(getAllByText(tabs.GAMES.label)[0]);
    expect(getByTestId(`tabpanel--${tabs.GAMES.id}`)).toBeTruthy();
  });
});
