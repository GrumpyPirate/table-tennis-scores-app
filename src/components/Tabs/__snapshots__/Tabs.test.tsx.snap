// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components/Tabs should render as expected, without crashing 1`] = `
<nav>
  <ul
    class="sc-kfzAmx fHNeVM"
  >
    <li
      class="sc-bBXqnf loegsG"
    >
      <button
        class="sc-fKFyDc bkYuUL"
        type="button"
      >
        Players
      </button>
    </li>
    <li
      class="sc-bBXqnf loegsG"
    >
      <button
        class="sc-fKFyDc jGhKL"
        type="button"
      >
        Games
      </button>
    </li>
  </ul>
</nav>
`;
